src/dialog/provider.js

```diff
@@ -18,107 +18,72 @@ export default class FlowProvider extends EventEmitter2 {
-
-    const flows = []
+    const flowFiles = await this.ghostManager.directoryListing(this.flowsDir, '.flow.json')
 
-    flowFiles.forEach(file => {
-      const filePath = path.resolve(flowsDir, './' + file)
-      const flow = JSON.parse(fs.readFileSync(filePath))
+    const flows = await Promise.all(
+      flowFiles.map(async name => {
+        const uiFileName = name.replace(/\.flow/g, '.ui')
+        const flow = JSON.parse(await this.ghostManager.readFile(this.flowsDir, name))
 
-      flow.name = file // e.g. 'login.flow.json' or 'shapes/circle.flow.json'
-
-      const schemaError = validateFlowSchema(flow)
-      if (schemaError) {
-        return this.logger.warn(schemaError)
-      }
-
-      const uiEqPath = path.resolve(flowsDir, './' + file.replace(/\.flow/g, '.ui'))
-      let uiEq = {}
+        const schemaError = validateFlowSchema(flow)
+        if (!flow || schemaError) {
+          return flow ? this.logger.warn(schemaError) : null
+        }
 
-      if (fs.existsSync(uiEqPath)) {
-        uiEq = JSON.parse(fs.readFileSync(uiEqPath))
-      }
+        const uiEq = JSON.parse(await this.ghostManager.readFile(this.flowsDir, uiFileName))
 
-      Object.assign(flow, { links: uiEq.links })
+        Object.assign(flow, { links: uiEq.links })
 
-      // Take position from UI files or create default position
-      const unplacedNodes = []
-      flow.nodes.forEach(node => {
-        const uiNode = _.find(uiEq.nodes, { id: node.id }) || {}
+        // Take position from UI files or create default position
+        flow.nodes.forEach(node => {
+          const uiNode = _.find(uiEq.nodes, { id: node.id }) || {}
+          Object.assign(node, uiNode.position)
+        })
 
-        Object.assign(node, uiNode.position)
+        const unplacedY = (_.maxBy(flow.nodes, 'y') || { y: 0 }).y + 250
+        flow.nodes.filter(node => _.isNil(node.x) || _.isNil(node.y)).forEach((node, i) => {
+          node.y = unplacedY
+          node.x = 50 + i * 250
+        })
 
-        if (_.isNil(node.x) || _.isNil(node.y)) {
-          unplacedNodes.push(node)
+        return {
+          name,
+          location: name,
+          nodes: _.filter(flow.nodes, node => !!node),
+          ..._.pick(flow, 'version', 'catchAll', 'startNode', 'links', 'skillData')
         }
       })
+    )
 
-      const unplacedY = (_.maxBy(flow.nodes, 'y') || { y: 0 }).y + 250
-      let unplacedX = 50
-
-      unplacedNodes.forEach(node => {
-        node.y = unplacedY
-        node.x = unplacedX
-        unplacedX += 250
-      })
-
-      return flows.push({
-        location: file,
-        version: flow.version,
-        name: flow.name,
-        nodes: _.filter(flow.nodes, node => !!node),
-        catchAll: flow.catchAll,
-        startNode: flow.startNode,
-        links: flow.links,
-        skillData: flow.skillData
-      })
-    })
-
-    return flows
+    return flows.filter(flow => Boolean(flow))
```

Comment:
NIT but can as well be `.filter(Boolean)`

Reply:
Good point!